codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,297 @@
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                // TODO: enable when syncing to master
+                                /*&& http2Handler.connection().local().isValidStreamId(stream.id())*/) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+                            if (streamInfo != null) {
+                                goAway.retain();
+                                streamInfo.childChannel.pipeline().fireUserEventTriggered(goAway);
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Http2Exception ex) {
+                ctx.invoker().invokeExceptionCaught(ctx, ex);
+            }
+            // TODO: who releases the original reference to debugData? Does this need a retain?
+            ctx.fireUserEventTriggered(goAway);
+        }
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            // Use a user event in order to circumvent read queue.
+            streamInfo.childChannel.pipeline().fireUserEventTriggered(
+                new Http2ResetMessage(errorCode).setStreamId(streamId));
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId,
+                Http2Headers headers, int streamDependency, short weight, boolean
+                exclusive, int padding, boolean endStream) throws Http2Exception {
+            onHeadersRead(ctx, streamId, headers, padding, endStream);
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                int padding, boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            fireChildReadAndRegister(streamInfo,
+                new Http2HeadersMessage(headers, endOfStream, padding).setStreamId(streamId));
+        }
+
+        @Override
+        public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            // TODO: Determine whether retain is needed.
```

Comment:
I think the `payload.skipBytes(payload.readableBytes());` is a noop except if there is padding, because readSlice() actually consumes from `payload`, advancing the `readerIndex`. Also, `skipBytes()` itself only advances the `readerIndex`, without freeing the backing memory (even in the case of `CompositeByteBuf`). The created slice doesn't retain() the parent automatically.

It seems [ByteToMessageDecoder](https://github.com/netty/netty/blob/b122faea6823d7daddfecf3e414237cf6ece28f9/codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java#L238) is releasing the buffer. I need a retain here.

It would be _really_ nice if lifecycles were documented for passed `ByteBuf`s. But that's a larger Netty issue.